### ISSUE TITLE
Bug: discrepancy between `onChange` and `addEventListener('change',...)` in checkboxes

### ISSUE THREAD

**Issue by marcellomontemagno** [facebook/react#18704]
I noticed that for controlled checkboxes there is a discrepancy between the checked value returned within an `onChange` callback and a callback added with `addEventListener` regardless of the `capture` flag set to `true`.

You can reproduce the issue in the following codesandbox by clicking on the checkbox and reading the logs.

https://codesandbox.io/s/reactcheckbox-ch2ho?file=/src/App.js

## The current behavior

`event.target.value` is `false` in the `onChange` callback and `true` in the `addEventListener` callback

## The expected behavior

`event.target.value` should have the same value in both callbacks, probably `true` since the checkbox is controlled and stuck with the checked value in the example

## Further info

I'm using chrome but to make it more interesting it seems that the value returned in the `addEventListener` callback is different if the code runs in ie11 which was the reason I started digging into all of this.

The callbacks seem to behave better on uncontrolled checkboxes but this behavior can be very confusing.

**Comment by facufrast** [comment]
Try change the 'true' value in line 23 with 'null'

**Comment by 7kms** [comment]
It is't discrepancy. React captured the inputchange event, then fire the `onChange` callback, then render the dom. 
After render, you checkbox is always checked because you set it a true value.
`addEventListener` is registered after the dom is rendered, the callback is always behind the `onChange` on. So the value is always true in the `addEventListener` callback.

**Comment by vinodsai-a** [comment]
Hi @bvaughn, Shall I work on this? (I'm new to opensource contribution)

**Comment by rachelnabors** [comment]
So it's not exactly a controlled component because you're not using state for un/checking the checkbox. This is expected behavior in this case: 
1) Input changes
2) `onChange` runs with the new value
3) Because you have `checked={true}` but didn't set state, React reverts to the component's original value of `true`
4) your native listener fires

And that describes the behavior here!

Here's an example of a controlled component that doesn't have this problem: https://codesandbox.io/s/reactcheckbox-forked-8wsl8?file=/src/App.js

Alternatively, you can make an uncontrolled component that also doesn't have this problem using [`defaultChecked`](https://reactjs.org/docs/uncontrolled-components.html#default-values) as in this example: https://codesandbox.io/s/reactcheckbox-forked-45xeq?file=/src/App.js

In short: if you're using a controlled component, make sure you `useState` to _really_ control your component ;)
